simuvex/plugins/symbolic_memory.py

```diff
@@ -649,132 +650,228 @@ def _store(self, req):
         l.debug("Doing a store...")
         req._adjust_condition(self.state)
 
-        if req.size is not None and self.state.se.symbolic(req.size) and options.AVOID_MULTIVALUED_WRITES in self.state.options:
-            return req
+        max_bytes = req.data.length/8
+
+        if req.size is None:
+            req.size = claripy.BVV(max_bytes, req.data.length)
+
+        if self.state.solver.symbolic(req.size):
+            if options.AVOID_MULTIVALUED_WRITES in self.state.options:
+                return req
+            if options.CONCRETIZE_SYMBOLIC_WRITE_SIZES in self.state.options:
+                new_size = self.state.solver.any_int(req.size)
+                req.constraints.append(req.size == new_size)
+                req.size = new_size
 
-        if self.state.se.symbolic(req.addr) and options.AVOID_MULTIVALUED_WRITES in self.state.options:
+        if self.state.solver.symbolic(req.addr) and options.AVOID_MULTIVALUED_WRITES in self.state.options:
             return req
 
-        if req.size is not None and self.state.se.symbolic(req.size) and options.CONCRETIZE_SYMBOLIC_WRITE_SIZES in self.state.options:
-            new_size = self.state.se.any_int(req.size)
-            req.constraints.append(req.size == new_size)
-            req.size = new_size
+        if not self.state.solver.symbolic(req.size) and self.state.solver.any_int(req.size) > req.data.length/8:
+            raise SimMemoryError("Not enough data for requested storage size (size: {}, data: {})".format(req.size, req.data))
 
-        max_bytes = len(req.data)/8
+        req.constraints += [self.state.solver.ULE(req.size, max_bytes)]
+
+        condition = req.condition if req.condition is not None else claripy.BoolV(True)
 
         #
         # First, resolve the addresses
         #
 
         try:
-            req.actual_addresses = self.concretize_write_addr(req.addr)
+            req.actual_addresses = sorted(self.concretize_write_addr(req.addr))
```

Comment:
Sorting is necessary for the `symbolic address` and `symbolic address+size` writes, where we have to calculate overlaps, can move the sorting into the particular cases though?